### PR TITLE
Keep information about CSE defs on assignment nodes

### LINKED ISSUE #64883: Assertion failed '!"Write to unaliased local overlaps outstanding read"' during 'Rationalize IR'
### Description

The following program hits a JIT assert for arm64 (also repros with altjit):

### Reproduction Steps

```csharp
// Generated by Fuzzlyn v1.5 on 2022-02-06 15:53:33
// Run on Arm64 Windows
// Seed: 7637938960038665944
// Reduced from 168.4 KiB to 0.2 KiB in 00:05:13
// Hits JIT assert in Release:
// Assertion failed '!"Write to unaliased local overlaps outstanding read"' in 'Program:Main(Fuzzlyn.ExecutionServer.IRuntime)' during 'Rationalize IR' (IL size 26)
// 
//     File: D:\a\_work\1\s\src\coreclr\jit\lir.cpp Line: 1397
// 
using System.IO;
using System.Reflection;
using System.Runtime.Loader;

public class Program
{
    public static uint s_29;
    public static void Main()
    {
        ServerAssemblyLoadContext alc = new ServerAssemblyLoadContext();
        Assembly asm = alc.LoadFromStream(new MemoryStream(File.ReadAllBytes(Assembly.GetExecutingAssembly().Location)));
        MethodInfo mi = asm.GetType("Program").GetMethod("MainT");
        mi.Invoke(null, new object[] { });
    }

    public static void MainT()
    {
        long vr7 = 4447329742151181917L;
        vr7 /= (vr7 ^ s_29);
        uint vr6 = s_29;
    }
}

public class ServerAssemblyLoadContext : AssemblyLoadContext
{
    public ServerAssemblyLoadContext() : base(true)
    {
    }
}
```

The following tree:
```scala
STMT00001 ( ??? ... ??? )
N023 ( 50, 45) [000015] -ACXG-------              ▌  COMMA     void   <l:$3c1, c:$3c0>
N021 ( 50, 45) [000013] -ACXG-------              ├──▌  DIV       long   <l:$34d, c:$34c>
N001 (  1,  1) [000040] ------------              │  ├──▌  LCL_VAR   long   V03 cse1         u:1 $142
N020 ( 29, 41) [000012] -ACXG-------              │  └──▌  XOR       long   <l:$347, c:$346>
N014 ( 23, 23) [000011] -ACXG----U--              │     ├──▌  CAST      long <- ulong <- uint <l:$341, c:$343>
N013 ( 22, 21) [000034] -ACXG-------              │     │  └──▌  COMMA     int    <l:$2c4, c:$2c3>
N011 ( 21, 20) [000032] -ACXG---R---              │     │     ├──▌  ASG       int    $VN.Void
N010 (  1,  1) [000031] D------N----              │     │     │  ├──▌  LCL_VAR   int    V02 cse0         d:1 <l:$2c4, c:$2c3>
N009 ( 21, 20) [000010] --CXG-------              │     │     │  └──▌  IND       int    <l:$2c4, c:$2c3>
N008 ( 20, 21) [000009] --CXG--N----              │     │     │     └──▌  ADD       byref  $203
N006 ( 18, 18) [000007] H-CXG-------              │     │     │        ├──▌  CALL help byref  HELPER.CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE_DYNAMICCLASS $201
N004 (  3, 12) [000005] ------------ arg0 in x0   │     │     │        │  ├──▌  CNS_INT   long   0x7ffed6e1d0b0 $140
N005 (  1,  2) [000006] ------------ arg1 in x1   │     │     │        │  └──▌  CNS_INT   int    0 $41
N007 (  1,  2) [000008] ------------              │     │     │        └──▌  CNS_INT   long   8 Fseq[s_29] $141
N012 (  1,  1) [000033] ------------              │     │     └──▌  LCL_VAR   int    V02 cse0         u:1 <l:$2c4, c:$2c3>
N019 (  5, 17) [000039] -A----------              │     └──▌  COMMA     long   $142
N017 (  4, 16) [000037] -A------R---              │        ├──▌  ASG       long   $VN.Void
N016 (  1,  1) [000036] D------N----              │        │  ├──▌  LCL_VAR   long   V03 cse1         d:1 $142
N015 (  4, 16) [000004] ------------              │        │  └──▌  CNS_INT   long   0x3db816d667f3665d $142
N018 (  1,  1) [000038] ------------              │        └──▌  LCL_VAR   long   V03 cse1         u:1 $142
N022 (  0,  0) [000014] ------------              └──▌  NOP       void   $380
```

Is rationalized into:
```scala
read:  N001 (  1,  1) [000040] ------------        t40 =    LCL_VAR   long   V03 cse1         u:1 $142
       N004 (  3, 12) [000005] ------------         t5 =    CNS_INT   long   0x7ffed6e1d0b0 $140
       N005 (  1,  2) [000006] ------------         t6 =    CNS_INT   int    0 $41
                                                         ┌──▌  t5     long   arg0 in x0
                                                         ├──▌  t6     int    arg1 in x1
       N006 ( 18, 18) [000007] H-CXG-------         t7 = ▌  CALL help byref  HELPER.CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE_DYNAMICCLASS $201
       N007 (  1,  2) [000008] ------------         t8 =    CNS_INT   long   8 Fseq[s_29] $141
                                                         ┌──▌  t7     byref
                                                         ├──▌  t8     long
       N008 ( 20, 21) [000009] ---XG--N----         t9 = ▌  ADD       byref  $203
                                                         ┌──▌  t9     byref
       N009 ( 21, 20) [000010] ---XG-------        t10 = ▌  IND       int    <l:$2c4, c:$2c3>
                                                         ┌──▌  t10    int
       N011 ( 21, 20) [000032] DA-XG-------              ▌  STORE_LCL_VAR int    V02 cse0         d:1
       N012 (  1,  1) [000033] ------------        t33 =    LCL_VAR   int    V02 cse0         u:1 <l:$2c4, c:$2c3>
                                                         ┌──▌  t33    int
       N014 ( 23, 23) [000011] ---XG----U--        t11 = ▌  CAST      long <- ulong <- uint <l:$341, c:$343>
       N015 (  4, 16) [000004] ------------         t4 =    CNS_INT   long   0x3db816d667f3665d $142
                                                         ┌──▌  t4     long
write: N017 (  4, 16) [000037] DA----------              ▌  STORE_LCL_VAR long   V03 cse1         d:1
       N018 (  1,  1) [000038] ------------        t38 =    LCL_VAR   long   V03 cse1         u:1 $142
                                                         ┌──▌  t11    long
                                                         ├──▌  t38    long
       N020 ( 29, 41) [000012] ---XG-------        t12 = ▌  XOR       long   <l:$347, c:$346>
                                                         ┌──▌  t40    long
                                                         ├──▌  t12    long
user:  N021 ( 50, 45) [000013] ---XG-------        t13 = ▌  DIV       long   <l:$34d, c:$34c>
```

### Expected behavior

Not JIT assert hit.

### Actual behavior

JIT assert hit.

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

cc @dotnet/jit-contrib

### PR DESCRIPTION
**Pull request by jakobbotsch**
Reuse the gtCSEnum field on assignments to signal "this is a performed
CSE def". gtCanSwapOrder uses the CSE def information to reason about
evaluation order, so without keeping this information we could risk
substitution a CSE def, followed by reordering operands, and then
finally substituting a CSE use.

Fix #64883

I also added some nicer output for when the assert is hit.

[No diffs.](https://dev.azure.com/dnceng/public/_build/results?buildId=1599230&view=ms.vss-build-web.run-extensions-tab)

### CODE REVIEW

**Comment by jakobbotsch** [comment]
cc @dotnet/jit-contrib

**Comment by jakobbotsch** [comment]
I will run outerloop to verify that the test that loads an assembly from file does not break any legs there.

File: src/tests/JIT/Regression/JitBlue/Runtime_64883/Runtime_64883.csproj [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <Optimize>True</Optimize>
```
Comment (BruceForstall) [inline src/tests/JIT/Regression/JitBlue/Runtime_64883/Runtime_64883.csproj:4]: I wonder if this should be Pri-1; seems rare, and maybe the "load assembly" is relatively slow?

File: src/coreclr/jit/optcse.cpp [hunk 2]
```diff
@@ -3230,6 +3207,14 @@ class CSE_Heuristic
                 // Backpatch the SSA def, if we're putting this CSE temp into ssa.
                 asg->AsOp()->gtOp1->AsLclVar()->SetSsaNum(cseSsaNum);
 
+                // Move the information about the CSE def to the assignment; it
+                // now indicates a completed CSE def instead of just a
+                // candidate. optCSE_canSwap uses this information to reason
+                // about evaluation order inbetween substitutions of CSE
```
Comment (BruceForstall) [inline src/coreclr/jit/optcse.cpp:3213]:
```suggestion
                // about evaluation order in between substitutions of CSE
```

**Review by BruceForstall — APPROVED** [review]

File: src/tests/JIT/Regression/JitBlue/Runtime_64883/Runtime_64883.csproj (hunk 1 above)
Comment (jakobbotsch) [reply src/tests/JIT/Regression/JitBlue/Runtime_64883/Runtime_64883.csproj:4]: Moved it to pri-1.

**Comment by jakobbotsch** [comment]
The outerloop failures do not look related to the newly added test, so will merge.

**Pull request merged**